### ISSUE TITLE
Typing text in the bash: existing text is overwritten, new text is not inserted

### ISSUE THREAD

**Issue by jepster** [Microsoft/BashOnWindows#1307]
* A brief description
When I type text into the bash, the existing text is overwritten. The new text is not inserted.

* Expected results
If I type new text, it should not overwrite the existing one. Outside of the bash, the result is the expected one.

* Your Windows build number
14393.0

* Steps / All commands required to reproduce the error from a brand new installation 
Just an open bash.

**Comment by zadjii-msft** [comment]
@jepster What PC are you using? More specifically, what graphics card (if any)?

Does this happen in cmd.exe or powershell? or is it only bash.exe?

Could you post a screenshot?

**Comment by diabolicallyrandom** [comment]
I am having the same / similar issue.

Basically, lets say I want to re-run the last entered command in a bash shell, but tweak some item in the command string. I hit the up arrow on my keyboard to repeat the last command, I arrow key over to where I want to make the change. 

Now when I use the delete key, it doesn't (visually) display the deleted characters being deleted. And it appears to overtype new characters on what is left there.

In actuality, it DOES change the text, as evidenced when pressing enter to submit the command, but its near impossible to actually do this reliably since the modification cannot be seen.

**Comment by jepster** [comment]
@zadjii-msft The issue appeared in the bash.exe. I've tried to use the bash.exe inside the PHPStorm IDE, but there I cannot use the arrow keys of my keyboard.

Here is some info about my PC. It is in german, but I guess that you are able to understand the system info from the context.
![PC Info](http://i.imgur.com/qbPVumx.png)

Here is info about my graphics card:
![Graphics Card](http://i.imgur.com/azRDXjU.png)

I have recorded a screencast of the behavior. You can get to the video by the following URL: https://drive.google.com/file/d/0B5vmNAcUsVUgSWtBS2pNTmFISGc/view

# How to reproduce this bug?
At the beginning, the text typing is ok. I have opened a file in VIM, edited it a bit. Then the behavior was still ok. Afterwards I was installing a program via "sudo apt install youtube-dl". Then I wanted to remove it from the system by "sudo apt purge youtube-dl". Afterwards the problem appeared. However - you can see this in my screencast - the issue disappeared after a while if typing.

What could this have to do with my graphics card? I doubt that the issue is related to the graphics card.

**Comment by diabolicallyrandom** [comment]
Just posting to confirm the screencast video Jepster posted is the identical behavior I am experiencing. I am experiencing it on multiple systems of different configuration, *including* a Hyper-V Server hosted Win10 VM, indicating (to me) there is no GPU issues at all.

I also experience similar behavior in that it does not occur until after the bash shell has been open for some indeterminate amount of time.

**Comment by zadjii-msft** [comment]
@jepster @diabolicallyrandom Hey so we may have found a hint in figuring this out.

Does this happen specifically after using apt to install something? Or can this happen in a fresh conhost.exe instance without really doing all that much?

I believe there's some program somewhere in the stack that's emitting something to cause conhost to get in a bad state, if only occasionally.

**Comment by diabolicallyrandom** [comment]
For my part, I have always done something prior to it occurring of course. I am typically using SSH to connect to my Ubuntu server, or my OpenWRT router. However I also had it occur after a session in which I used apt to install image magic and work with some files on my mounted windows drives.

If I can fine the time I will try and replicate the behavior without doing anything in the shell, because I swear I had that happen too.

Just to confirm, this happens on command prompt bash, PowerShell bash, and launching the bash shortcut directly. And it usually (although sometimes) doesn't happen while I'm actively using it, but often it's when I use it and walk away for some time and come back.

Would there be a way I could provide some kind of debug logging that might help?

**Comment by zadjii-msft** [comment]
If you could help get a reliable repro, that'd be huge. This has proven to be wildly difficult to get a consistent repro from. Just now, I found that on a small enough window with a sufficiently large screen buffer, alternating `sudo apt install nmap` and `sudo apt remove nmap` can cause some input line corruption. However, there's a good chance I'll come back to my desk on Monday and it'll be gone
<hr>
>Just to confirm, this happens on command prompt bash, PowerShell bash, and launching the bash shortcut directly. 

For the record, all of those are all the same bash.exe running in conhost.exe. If it happened in plain old cmd.exe (no bash involved), now that certainly would be a bigger issue, but seeing how it's only a bash issue we can probably assume that there's some VT sequence that's messing with the console.

**Comment by jepster** [comment]
Guys, I was typing a lot in the bash and I had this buggy behavior after I have run some PHP command line application. Then - in the hope that this was the issue causer - I have run the app again. But the Bash was behaving correct. Could it be, that there is any issue with the memory of the bash? Because the problem was never appearing directly but only after a while of working in the Bash. That would also describe why this issue happens with "apt". Because "apt" creates a lot of output on installing any application. Well, this is only "guessing". Though a reasonable indicator.

**Comment by lilred** [comment]
I'm consistently seeing this problem on 14965. Vim is also affected. Demo: https://www.youtube.com/watch?v=eDd9A8YJCH8

**Comment by pszafer** [comment]
I have the same problem as @lilred 
SSH via bash on windows is almost unusable with that bug...

**Comment by zadjii-msft** [comment]
I eventually was able to get a consistent repro on this bug, so I'm opening MSFT:9871791 to investigate further. This one's probably the highest priority bug ATM.

**Comment by jepster** [comment]
I am glad to read that. Thanks for your quick feedback.

**Comment by jepster** [comment]
@zadjii-msft: Any news regarding the bug? Where can we see the state of "MSFT:9871791"? I guess this is a Microsoft bug tracker which is hidden for the public.

**Comment by zadjii-msft** [comment]
It is our internal tracking number. It's under investigation, but I've been out for the holidays, so nothing new to report.

**Comment by jepster** [comment]
Any news? (:

**Comment by zadjii-msft** [comment]
@jepster Nothing yet. I think I got it pinned down in the code, but I don't have a fix tested yet.

**Comment by zadjii-msft** [comment]
So, breakdown of what's going on here:

`apt install` (NOT `apt-get install`) uses `'\x1b[1;{{height-1}}r'` to set the scrolling margins of the terminal to one less than the entire terminal while it's displaying output (including that neat little loading bar of theirs). They then use `'\x1b[1;{{height}}r'` to reset the margins back to the entire screen height, effectively removing them.

Typically, setting the scroll margins is only done inside the Alternate Buffer on a terminal, eg, within an application like vim, and not usually on the main buffer. Resetting the margins by setting to the entire terminal height is fine on the alt buffer, because it is exactly the size of the terminal window (it has no scrollback). 

However, conhost doesn't really have a traditional "scrollback", and if you set the margins to the entire terminal on the main buffer in conhost, then they would stay there, _even once you scrolled away_. This could create scenarios where certain text manipulations would be clipped to the old scroll margins (which were no longer on the screen). This would cause visual artifacts in the output buffer, even though the `bash` input line still logically knew the correct data.

I've put the fix in code review today. Thanks for all the help on this one everyone, pinning this one down was particularly challenging.

**Comment by jepster** [comment]
Thanks for your very detailed feedback! It's great that Linux gets better and better as a subsystem in Windows by your effort. Can you share with us how the process looks like, until we will receive your update on our Windows 10 systems? Do you have an idea how much time it will take more or less?

**Comment by zadjii-msft** [comment]
@jepster Basically, the process for a pretty trivial fix like this is as follows:
 *  it'll bounce around code review a bit (for such a small fix, this doesn't take too long)
 * it then gets checked in to our teams source branch.
 * It now needs about 3 weeks to get up to winmain. This time varies depending on how stable winmain is, or if our sister team's quality remains good, or hey, even a holiday break (like the one at the end of the month) can put this off by a week. 3 is what I've found to be the average. Sometimes all the pieces line up just wrong and this can be MUCH longer.
 * Once it's in winmain, it's usually the following week's insider build that has the actual fix in it. I believe it goes into the fast ring then, but I don't really know enough to know how long it takes to go from Fast -> Slow.
 * And it'll hit non-Insiders whenever the Creator's Update hits

**Comment by jepster** [comment]
@zadjii-msft: Thanks for your quick reply. Sounds like a straight forward process. 

I am an non-insider. Can you tell me, how much difference there is between the non-insider and the insider release?

**Comment by zadjii-msft** [comment]
@jepster Easier for me to just link this: http://www.windowscentral.com/all-you-need-know-about-windows-insider-rings

**Comment by diabolicallyrandom** [comment]
Just to verify, this fix should fix any instances of similar behavior, not just those triggered by apt, correct?

**Comment by zadjii-msft** [comment]
@diabolicallyrandom Theoretically yes. Turns out there was a few related bugs related to the scroll margins that I all just checked in. I'm marking this particular issue as "fixinbound" and I'll close once it gets to Insiders.

If it seems to resurface once those changes are in Insiders, then it's probably an unrelated bug causing those, and I'd love a new issue opened to track it, to help single out the root cause

**Comment by diabolicallyrandom** [comment]
OK Thanks!

**Comment by miniksa** [comment]
The fix for this has just been released with Insiders build 15014.

**Comment by asampal** [comment]
I'm seeing this with the latest bits from the CU on bash 4.3.48(1). Should it not have been fixed?

**Comment by zadjii-msft** [comment]
@asampal I'm gonna refer you to #1756, which is basically tracking this issue since creator's update. `apt install` causing the input line to act weird should be fixed in Creator's Update

**Comment by asampal** [comment]
@zadjii-msft , I'm not sure I follow - are you saying that #1756 (which describes problems on the input line) should be fixed for me? In any case, I commented there as well.

**Comment by JessePotterMusic** [comment]
I am having this issue to when I run the mongo shell in bash(babun). Is it the kind of thing that is just predetermined by the software you're using? Should I write to mongo and ask for a fix?

**Comment by zadjii-msft** [comment]
@JessePotterMusic If this is still happening to you, and you have some reliable repro steps, you should file a new issue here. That's the best way to get me to start investigating :)

**Comment by HarishMundluru1998** [comment]
It's still happening for me as well.

**Comment by tfinn81** [comment]
Also still having this issue. Surprised it hasn't been fixed yet.

**Comment by zadjii-msft** [comment]
> If this is still happening to you, and you have some reliable repro steps, you should file a new issue [here](https://github.com/microsoft/terminal). 

The issue this thread was tracking was fixed in 2017. If you're still seeing it, then it's likely a different bug, with similar symptoms.

**Comment by andrewnryan** [comment]
I realize this issue is closed, but wanted to post here for folks like me who still see similar symptoms and happen across this link

> 
> In a possibly related issue (#2713) , I found a [workaround](https://github.com/microsoft/WSL/issues/2713#issuecomment-530975323) that worked for me.

These work around items (Setting the text wrap checkbox and adjusting PS1) did not result in any noticeable difference in behavior, and text in remote sessions is still being overwritten. 

> 
> This stopped for me after switching to Debian from ubuntu on wsfl.

Can confirm this for me as well.  I tried Ubuntu18.04 LTS and Ubuntu 20.04 LTS (both direct from Microsoft Store) and both had the overwriting issue.  When I installed Debian, I see normal behavior where text is no longer being overwritten.

**Comment by mzyil** [comment]
> export TERM=vt100 on my ubuntu/debian
this removes color from my terminal, instead I set it to this value:
`export TERM=xterm-256color`
now it works fine. It was `xterm-color` previously, set in my decade-old .bashrc file that I carried around.
